node/src/components/consensus/era_supervisor.rs

```diff
@@ -181,68 +190,47 @@ where
             era_where_we_joined: current_era,
         };
 
-        let bonded_eras = era_supervisor.bonded_eras();
-        let era_ids: Vec<EraId> = era_supervisor
-            .iter_past(current_era, bonded_eras.saturating_mul(3))
-            .collect();
-
-        // Collect the information needed to initialize all recent eras.
-        // TODO: Merge these into one
-        let mut key_blocks = BTreeMap::new();
-        let mut booking_blocks = BTreeMap::new();
-        for era_id in era_ids {
-            let minimum_era = era_supervisor
-                .protocol_config
-                .last_emergency_restart
-                .unwrap_or_else(|| EraId::from(0u64));
-            let key_block_era_id = era_id.saturating_sub(1);
-            let block_header = storage
-                .read_switch_block_header_by_era_id(key_block_era_id)?
-                .ok_or_else(|| anyhow::Error::msg("No such switch block"))?;
-            key_blocks.insert(era_id, block_header);
-            let booking_block_era_id = era_id
-                .saturating_sub(era_supervisor.protocol_config.auction_delay)
-                .saturating_sub(1);
-            let booking_block_hash = if booking_block_era_id < minimum_era {
-                Default::default()
-            } else {
-                storage
-                    .read_switch_block_header_by_era_id(booking_block_era_id)?
-                    .ok_or_else(|| anyhow::Error::msg("No such switch block"))?
-                    .hash()
-            };
-            booking_blocks.insert(era_id, booking_block_hash);
+        // Collect the information needed to initialize all active eras.
+        //
+        // We need to initialize current_era, current_era - 1 and (evidence-only) current_era - 2.
+        // To initialize an era, all switch blocks between its booking block and its key block are
+        // required. The booking block for era N is in N - auction_delay - 1, and the key block in
+        // N - 1. So we need all switch blocks between:
+        // (including) current_era - 2 - auction_delay - 1 and (excluding) current_era.
+        // However, we never use any block from before the last activation point.
+        //
+        // Example: If auction_delay is 1, to initialize era N we need the switch blocks from era N
+        // and N - 1. If current_era is 10, we will initialize eras 10, 9 and 8. So we need the
+        // switch blocks from eras 9, 8, 7 and 6.
```

Comment:
This is a really good explanation that will definitely make it easier for people to follow the code. Thanks 👍